### PR TITLE
feat: add `ndarray/map`

### PR DESCRIPTION
**Pull request by headlessNode**
> What is the purpose of this pull request?

This pull request:

-   add top-level implementation of `map` function for `@stdlib/ndarray`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2656 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/ndarray/map/examples/index.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2023 The Stdlib Authors.
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/examples/index.js]:
```suggestion
* Copyright (c) 2024 The Stdlib Authors.
```

File: lib/node_modules/@stdlib/ndarray/map/docs/repl.txt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    // Define the options:
+    > var opts = { 'dtype': 'float32' };
+
+    // Using ndarray...
+    > var x = {{alias:@stdlib/ndarray/ctor}}( dtype, buffer, shape, strides, offset, order );
+    > var y = {{alias}}( x, opts, f );
+    > y.data
+    <Float32Array>[ 10.0, 20.0, 30.0, 40.0 ]
+
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/docs/repl.txt]:
```suggestion
```

File: lib/node_modules/@stdlib/ndarray/map/lib/main.js [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+*
+* @param {ndarray} x - input ndarray
+* @param {Options} [options] - function options
+* @param {string} [options.dtype] - output array data type (overrides the input array's inferred data type)
+* @param {Callback} fcn - callback function
+* @param {*} [thisArg] - callback execution context
+* @throws {TypeError} first argument must have a recognized data type
+* @throws {TypeError} callback argument must be a function
+* @throws {TypeError} options argument must be an object
+* @throws {TypeError} `dtype` option must be a supported ndarray data type
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/lib/main.js]:
```suggestion
```

File: lib/node_modules/@stdlib/ndarray/map/test/test.js [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	function badValue( value ) {
+		return function badValue() {
+			map( value, scale );
+		};
+	}
+});
+
+tape( 'the function throws an error if the second argument is not a function when options are not provided', function test( t ) {
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/test/test.js]:
```suggestion
tape( 'the function throws an error if the second argument is not a function', function test( t ) {
```

File: lib/node_modules/@stdlib/ndarray/map/test/test.js [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+	t.end();
+
+	function badValue( value ) {
+		return function badValue() {
+			map( x, value );
+		};
+	}
+});
+
+tape( 'the function throws an error if third argument is not a function when options are provided', function test( t ) {
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/test/test.js]:
```suggestion
tape( 'the function throws an error if third argument is not a function (options)', function test( t ) {
```

File: lib/node_modules/@stdlib/ndarray/map/test/test.js [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	function badValue( value ) {
+		return function badValue() {
+			map( x, value, scale );
+		};
+	}
+});
+
+tape( 'the function throws an error if provided an options argument which does not have a `dtype` property', function test( t ) {
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/test/test.js]:
```suggestion
tape( 'the function throws an error if the provided options argument does not have a `dtype` property', function test( t ) {
```

File: lib/node_modules/@stdlib/ndarray/map/test/test.js [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	function badValue( value ) {
+		return function badValue() {
+			map( x, value, scale );
+		};
+	}
+});
+
+tape( 'the function applies a callback to each indexed element in an input 3-dimensional ndarray and returns an output 3-dimensional ndarray with mapped values (row-major, contiguous)', function test( t ) {
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/test/test.js]:
```suggestion
tape( 'the function applies a callback to each indexed element in an input ndarray and returns an output ndarray with mapped values (row-major, contiguous)', function test( t ) {
```

File: lib/node_modules/@stdlib/ndarray/map/test/test.js [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+	t.strictEqual( isSameFloat64Array( y.data, expected ), true, 'returns expected value' );
+
+	t.end();
+
+	function scale( z ) {
+		return z * 10.0;
+	}
+});
+
+tape( 'the function applies a callback to each indexed element in an input 3-dimensional ndarray and returns an output 3-dimensional ndarray with mapped values, using the provided options (row-major, contiguous)', function test( t ) {
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/test/test.js]:
```suggestion
tape( 'the function applies a callback to each indexed element in an input ndarray and returns an output ndarray with mapped values(row-major, contiguous, options)', function test( t ) {
```

File: lib/node_modules/@stdlib/ndarray/map/test/test.js [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+		10.0
+	]);
+	t.strictEqual( isSameComplex128Array( y.data, expected ), true, 'returns expected value' );
+
+	t.end();
+
+	function scale( z ) {
+		return new Complex128( z*10.0, z*10.0 );
+	}
+});
```
Comment (headlessNode) [inline lib/node_modules/@stdlib/ndarray/map/test/test.js]:
```suggestion
```

**Review by kgryte — APPROVED** [review]
Thanks for working on this, @headlessNode. Did a bit of clean-up. This now LGTM.

**Pull request merged**